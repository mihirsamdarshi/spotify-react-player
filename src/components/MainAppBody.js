import React, { useEffect, useState } from 'react';
import { Grid, Modal, Paper, Typography } from '@material-ui/core';
import Playlists from './Playlists';
import Songs from './Songs';
import { fetchPlaylistTracks, fetchUserPlaylists } from '../scripts/api';
import useWindowDimensions from '../scripts/WindowDimensions';
import { GetNowPlayingDispatch, GetSongListDispatch } from '../scripts/callbacks';
import '../stylesheets/MainAppBody.scss';
import NowPlaying from './NowPlaying';


const MainAppBody = props => {
    const [showPlaylists, setShowPlaylists] = useState(false);
    const [showSongs, setShowSongs] = useState(false);
    const [showNowPlaying, setShowNowPlaying] = useState(false);
    const [playlistList, setPlaylistList] = useState('');
    const [songList, setSongList] = useState('');
    const [songPlaying, setSongPlaying] = useState('')
    const [errorString, setError] = useState(null);

    const { height, width } = useWindowDimensions();
    // subtract height of header + padding to make web app single page
    const heightWidthStyle = {
        height: height - 84,
        width: width - 84,
    };

    // Getter Functions
    const getPlaylists = async (token) => {
        setError(null);
        try {
            const result = await fetchUserPlaylists(token);
            setPlaylistList(result.items);
            setShowPlaylists(true);
        } catch (error) {
            setError('Sorry, but something went wrong.');
        }
    };

    const getSongs = async (link, token) => {
        console.log(link);
        console.log(token);
        setError(null);
        try {
            const result = await fetchPlaylistTracks(link, token);
            setSongList(result.items);
            setShowSongs(true);
        } catch (error) {
            setError('Sorry, but something went wrong.');
        }
    };

    const getNowPlaying = (value) => {
        setSongPlaying(value);
        setShowNowPlaying(true);
    };

    // Display Functions
    const numPlaylists = () => (
        <div className="numTracksWrapper">
            <Typography className="playlistsAvailable">
                {`${playlistList.length} playlists available`}
            </Typography>
        </div>
    );

    const numSongs = () => (
        <div className="numTracksWrapper">
            <Typography className="playlistsAvailable">
                {`${songList.length} songs available`}
            </Typography>
        </div>
    );

    const displayPlaylists = () => (
        <GetSongListDispatch.Provider value={getSongs}>
            <Playlists playlists={playlistList} token={props.token} />
        </GetSongListDispatch.Provider>
    );

    const displayError = () => (
        <Modal open>
            <p>{errorString}</p>
        </Modal>
    );

    const displaySongs = () => (
        <Paper className="paper" style={blackOverride}>
            <GetNowPlayingDispatch.Provider value={getNowPlaying}>
                <Songs songs={songList} />
            </GetNowPlayingDispatch.Provider>
        </Paper>
    );

    const displayNowPlaying = () => (
        <div className="paper">
            <NowPlaying nowPlaying={songPlaying} />
        </div>
    );

    // OnLoad Functions
    useEffect(() => {
        getPlaylists(props.token);
    }, [props]); //es

    // Handler Functions
    const handleSongListClose = () => {
        if (showSongs && !showNowPlaying) {
            setShowSongs(false);
        }
    };

    // Style Override Object
    const blackOverride = {
        backgroundColor: '#EEE',
    };

    return (
        <div className="root">
            <Grid container className="gridContainer">
                <Grid item xs={4} style={heightWidthStyle}>
                    <Paper className="paper" style={blackOverride}>
                        {showPlaylists && !errorString ? displayPlaylists() : displayError()}
                    </Paper>
                </Grid>
                <Grid item xs={4} style={heightWidthStyle}>
                    {showSongs ? displaySongs() : numPlaylists()}
                </Grid>
                <Grid item xs={4} style={heightWidthStyle} onClick={handleSongListClose}>
                    {showNowPlaying ? displayNowPlaying() : null}
                    {!showNowPlaying && showSongs ? numSongs() : null}
                </Grid>
            </Grid>
        </div>
    );
};

export default MainAppBody;
