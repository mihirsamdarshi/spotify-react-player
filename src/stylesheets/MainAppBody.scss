.root {
  flex-grow: 1;
}

.paper {
  padding: 15px;
  text-align: center;
  overflow: auto;
  height: 97%;
  margin-left: 1vw;
  margin-right: 1vw;
}

.paper.nowPlaying {
  height: 30vh;
}