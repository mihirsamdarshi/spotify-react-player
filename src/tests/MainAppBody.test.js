import React from 'react';
import ReactDOM from 'react-dom';
import TestRenderer from 'react-test-renderer';
import ReactTestUtils from 'react-dom/test-utils';
import { Grid, Paper, Typography } from '@material-ui/core';
import MainAppBody from '../components/MainAppBody';

const { act } = TestRenderer;


// This test suite uses a distinct testing technique called _snapshot testing_. Go take
// a peek at the code then come back here for more commentary.
//
// Note how, with snapshot testing, you are truly dependent on that descriptive text.
// The enforcement is in the snapshot match, not a condition that is in the test code.
// This is where snapshot testing differs from traditional test-driven development:
// _It assumes that the code works initially._ This actually does line up fairly well
// with user interface development, because it tends to be easier to just “eyeball” a
// user interface first rather than write tests against it.
//
// It takes some adjustment to start “trusting” a snapshot test, just as it takes some
// adjustment to trust version control. If you want to manually check whether a snapshot
// is truly in the state that you want it to be, you can always look at the .snap file
// within the __snapshots__ folder.
//
// Handy reference:
// https://semaphoreci.com/community/tutorials/snapshot-testing-react-components-with-jest
//
let container;

beforeEach(() => {
    container = document.createElement('div');
    document.body.appendChild(container);
});

afterEach(() => {
    document.body.removeChild(container);
    container = null;
});

it('renders the Main App Body component', async () => {
    await ReactTestUtils.act(async () => {
        ReactDOM.render(<MainAppBody />, container);
    });
});

it('should start with a list of playlists', async () => {
    await act(async () => {
        const component = await TestRenderer.create(<MainAppBody />);
        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });
});

describe('the MainAppBody', () => {
    it('should have four MUI Grid elements', async () => {
        await act(async () => {
            const component = await TestRenderer.create(<MainAppBody />);
            const componentCount = component.root.findAllByType(Grid).length;
            expect(componentCount).toBe(4);
        });
    });

    it('should have no MUI Paper elements', async () => {
        await act(async () => {
            const component = await TestRenderer.create(<MainAppBody />);
            const componentCount = component.root.findAllByType(Paper).length;
            expect(componentCount).toBe(0);
        });
    });

    it('should have one MUI Typography element', async () => {
        await act(async () => {
            const component = await TestRenderer.create(<MainAppBody />);
            const componentCount = component.root.findAllByType(Typography).length;
            expect(componentCount).toBe(1);
        });
    });

    it('should not have rendered the playlist component yet', async () => {
        await ReactTestUtils.act(async () => {
            ReactDOM.render(<MainAppBody />, container);
            const playlistComponent = container.querySelector('.playlistComponent');
            expect(playlistComponent !== null).toBe(false);
        });
    });

    it('should not have rendered the songs component', async () => {
        await ReactTestUtils.act(async () => {
            ReactDOM.render(<MainAppBody />, container);
            const songsComponent = container.querySelector('.songsComponent');
            expect(songsComponent !== null).toBe(false);
        });
    });

    it('should not have rendered the now playing component', async () => {
        await ReactTestUtils.act(async () => {
            ReactDOM.render(<MainAppBody />, container);
            const nowPlayingComponent = container.querySelector('.nowPlayingComponent');
            expect(nowPlayingComponent !== null).toBe(false);
        });
    });

    describe('upon clicking a Playlist', () => {
        it('should render a list of songs', async() => {
            await ReactTestUtils.act(async () => {
                ReactDOM.render(<MainAppBody />, container);
                const nowPlayingComponent = container.querySelector('.nowPlayingComponent');
                console.log(nowPlayingComponent);
            });
        });
    });
});
