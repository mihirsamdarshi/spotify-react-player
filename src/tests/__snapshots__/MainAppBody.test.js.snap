// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should start with a list of playlists 1`] = `
<div
  className="root"
>
  <div
    className="MuiGrid-root gridContainer MuiGrid-container"
  >
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4"
      style={
        Object {
          "height": 684,
          "width": 940,
        }
      }
    />
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4"
      style={
        Object {
          "height": 684,
          "width": 940,
        }
      }
    >
      <div
        className="numTracksWrapper"
      >
        <p
          className="MuiTypography-root playlistsAvailable MuiTypography-body1"
        >
          0 playlists available
        </p>
      </div>
    </div>
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4"
      onClick={[Function]}
      style={
        Object {
          "height": 684,
          "width": 940,
        }
      }
    />
  </div>
</div>
`;
